File: api/src/main/java/io/github/mh321productions/serverapi/util/logging/PixelLogHandler.java
package io.github.mh321productions.serverapi.util.logging;

import io.github.mh321productions.serverapi.util.formatting.StringFormatter;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.nio.charset.StandardCharsets;
import java.util.logging.Handler;
import java.util.logging.LogRecord;

public class PixelLogHandler extends Handler {
	
	private static final String format = "%s %s %s %s"; //Datum/Zeit Level Message Exception
	
	private BufferedWriter writer = null;
	
	public PixelLogHandler(File logFile) throws FileNotFoundException {
		writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(logFile), StandardCharsets.UTF_8));
	}

	@Override
	public void publish(LogRecord record) {
		if (writer != null) {
			String datum = StringFormatter.formatDateTime(record.getInstant());
			String level = record.getLevel().getName();
			String message = record.getMessage();
			String throwable = "";
			
			if (record.getThrown() != null) {
				StringWriter sw = new StringWriter();
				PrintWriter pw = new PrintWriter(sw);
				pw.println();
				record.getThrown().printStackTrace(pw);
				pw.close();
				throwable = sw.toString();
			}
			
			try {
				writer.write(String.format(format, datum, level, message, throwable));
				writer.newLine();
			} catch (IOException e) {
				e.printStackTrace();
			}
			
			flush();
		}
	}

	@Override
	public void flush() {
		try {
			writer.flush();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	@Override
	public void close() throws SecurityException {
		if (writer != null) {
			try {
				writer.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

}

File: src/main/kotlin/io/github/mh321productions/serverapi/util/logging/PixelLogHandler.kt
package io.github.mh321productions.serverapi.util.logging

import io.github.mh321productions.serverapi.util.formatting.StringFormatter.formatDateTime
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.logging.Handler
import java.util.logging.LogRecord

class PixelLogHandler(logFile: File) : Handler() {

    companion object {
        private const val format = "%s %s %s %s" //Datum/Zeit Level Message Exception
    }

    private val writer = BufferedWriter(OutputStreamWriter(FileOutputStream(logFile), StandardCharsets.UTF_8))

    override fun publish(record: LogRecord) {
        val datum = formatDateTime(record.instant)
        val level = record.level.name
        val message = record.message
        var throwable = ""

        if (record.thrown != null) {
            val sw = StringWriter()
            val pw = PrintWriter(sw)
            pw.println()
            record.thrown.printStackTrace(pw)
            pw.close()
            throwable = sw.toString()
        }

        try {
            writer.write(String.format(format, datum, level, message, throwable))
            writer.newLine()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        flush()
    }

    override fun flush() {
        try {
            writer.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Throws(SecurityException::class)
    override fun close() {
        try {
            writer.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
